lib/CustomPlugin.ts

```diff
@@ -0,0 +1,24 @@
+import { Router } from 'express'
+import MqttClient from './MqttClient'
+import { ModuleLogger } from './logger'
+import ZwaveClient from './ZwaveClient'
+
+export type PluginContext = {
+	zwave: ZwaveClient
+	mqtt: MqttClient
+	app: Router
+	logger: ModuleLogger
+}
+
+export interface CustomPlugin {
+	zwave: ZwaveClient
+	mqtt: MqttClient
+	app: Router
+	logger: ModuleLogger
+	name: string
+
+	// eslint-disable-next-line @typescript-eslint/no-misused-new
+	new (context: PluginContext): CustomPlugin
```

Comment:
This isn't correct. `CustomPlugin` defines the instance type, but the constructor lives on the prototype (the class itself). This definition makes it so that every plugin implementation literally needs a `new` function which is impossible to implement.

If you want to say a function can only accept a constructor of a certain type, this is the way to do it:
https://www.typescriptlang.org/play?#code/PTAEBEFcFtoT1AFzgBwKYGdQYPZIBZqgBGaANjgO4ECWAdgOaiU4BOA1hgFDLqgBalAIYA3NAGEyNNHUSgAvKABEg0RKkzESgNw9URALIBHRIknTZC5cdPnNOvXwBKOSIjSsrSl248PehjgAJpBkaAAyOAwMHl4GwaERUTGsDlwgtFgY+K5kQSREQnSg9O6sAGZCAMZoXGgAHihscqUelTWgAApkkAz04jiyDXIA3lygoABewmIAXAIz6haI46DQJojzNmYasqtCKCjzPmWrFNEeWwlhkResXAC+XHWNzSVDFdVE4pAYiDjQbq9eigYYyIJYIF9OgDIb1UarOhCaBoeZ-Vj0BirIKYRCsHBwAAUAEp5p18dAaBg0AAeEQ4GhBAB8j2eDSarDkAS6PWhsPRkCq-08ijoaGohKqg3c8LJvP60uGxIUTNAPz+AKh9F0z3KkDoQpog1AVVYaCE7i1dBpABVQfDwZD5TDBgKhWwmZLXXj5jbSWrfv9Ac7QGMJlK6H9QGa5KLxSbvaxCfqceV6Gh8kIsEU4MTdBMYwA6JEorzlHA4YhCVL50AZSD4UA4fAAGgKplidBwciqQjIZEyqyLOPRBJJtZjkFYxRjuieXCqZCzWAAYhWrSVoCgwijZFh1UGN2HQCWiIolA4JiO8WP-eSAVTafTGarjxNEPh8dQxdQAKKsfEkyUAw0A-YIT27TdtzQXd3CCQslDzVYngmaY1C8VQxDsWQlFALME0jRBa3WUw4g2bCtDwrAIz+WsDhQLwTj8KiCNos5klic94hCG4ONSFiaKI1kFyXDAsAAOW7ABBI9VgAfRCWAEHPBx51Nc1LWdQk1xwK08zrMAcHYBczQtNArUJSTEBk519IyDxAKAA